File: python/leetcode/binary_tree_postorder_traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    # @param root, a tree node
    # @return a list of integers

    def postorderTraversal(self, root):

        def helper(root):
            if root:
                if root.left:
                    for i in helper(root.left):
                        yield i
                if root.right:
                    for i in helper(root.right):
                        yield i
                yield root

        return [i.val for i in helper(root)]

File: java/src/Leetcode/BinaryTreePostorderTraversal.java
package Leetcode;

import Structs.TreeFactory;
import Structs.TreeNode;

import java.util.ArrayList;

/**
 * Created by xin on 3/6/14.
 */
public class BinaryTreePostorderTraversal {
    public ArrayList<Integer> postorderTraversal(TreeNode root) {
        ArrayList<Integer> res = new ArrayList<Integer>();
        if (root == null) return res;
        for (int i : postorderTraversal(root.left)) res.add(i);
        for (int i : postorderTraversal(root.right)) res.add(i);
        res.add(root.val);
        return res;
    }

    public static void main(String[] args) {
        TreeFactory f = new TreeFactory();
        int[] A = {1, 2, 3, 4, 5, 6, 7};
        //   1
        // 2  3
        //4 5 6 7
        TreeNode t = f.array2tree(A);
        BinaryTreePostorderTraversal s = new BinaryTreePostorderTraversal();
        for (int i : s.postorderTraversal(t)) {
            System.out.println(i);
        }
        // should be 4526731
    }
}
